### PR TITLE
hal/gl: Allow push constants trough emulation

### PR DESCRIPTION
**Pull request by JCapucho**
**Connections**
Depends on https://github.com/gfx-rs/naga/pull/1672 so bumps `naga` revision to pick it up

**Description**
Push constants are a great way to pass small amounts of data to the gpu between draw calls.

But maintaining two code paths for backends that support push constants and those who don't can be prohibitive to picking them.

So a good middle ground is to emulate push constants in the secondary backends, this is done by using a uniform buffer for push constants.

*Implementation notes*:

The buffer is always kept mapped and when a `set_push_constants` call is issued the mapping is written to.

Push constants stages are ignored.

The uniform buffer uses the binding 0 in the shader.

**TODOS**:

- [ ] The code currently uses the pointer returned from `glMapBufferRange` as a `*mut u32` for the slice, but according to https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glMapBufferRange.xhtml alignment is not guaranteed so it might need to be changed to a `*mut u8` and `set_push_constants` will need to convert the slice from `u32` to `u8`
- [ ] Another problem is synchronization, `glMapBufferRange` doesn't give much information but has an `GL_MAP_UNSYNCHRONIZED_BIT` so it leads me to believe that synchronization is implicit, but if someone who knows opengl better could confirm it would be helpful

**Testing**
I tested on a modified cube example
```patch
diff --git a/wgpu/examples/cube/main.rs b/wgpu/examples/cube/main.rs
index 60a49542..556ec979 100644
--- a/wgpu/examples/cube/main.rs
+++ b/wgpu/examples/cube/main.rs
@@ -124,10 +124,21 @@ impl Example {
 }
 
 impl framework::Example for Example {
+    fn required_features() -> wgt::Features {
+        wgt::Features::PUSH_CONSTANTS
+    }
+
     fn optional_features() -> wgt::Features {
         wgt::Features::POLYGON_MODE_LINE
     }
 
+    fn required_limits() -> wgpu::Limits {
+        wgpu::Limits {
+            max_push_constant_size: 64,
+            ..wgpu::Limits::downlevel_webgl2_defaults()
+        }
+    }
+
     fn init(
         config: &wgpu::SurfaceConfiguration,
         _adapter: &wgpu::Adapter,
@@ -179,7 +190,10 @@ impl framework::Example for Example {
         let pipeline_layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
             label: None,
             bind_group_layouts: &[&bind_group_layout],
-            push_constant_ranges: &[],
+            push_constant_ranges: &[wgpu::PushConstantRange {
+                stages: wgpu::ShaderStages::VERTEX,
+                range: 0..4,
+            }],
         });
 
         // Create the texture
@@ -380,6 +394,7 @@ impl framework::Example for Example {
             rpass.set_bind_group(0, &self.bind_group, &[]);
             rpass.set_index_buffer(self.index_buf.slice(..), wgpu::IndexFormat::Uint16);
             rpass.set_vertex_buffer(0, self.vertex_buf.slice(..));
+            rpass.set_push_constants(wgpu::ShaderStages::VERTEX, 0, &(0.2_f32).to_ne_bytes());
             rpass.pop_debug_group();
             rpass.insert_debug_marker("Draw!");
             rpass.draw_indexed(0..self.index_count as u32, 0, 0..1);
diff --git a/wgpu/examples/cube/shader.wgsl b/wgpu/examples/cube/shader.wgsl
index 4bd4305a..ba8423b4 100644
--- a/wgpu/examples/cube/shader.wgsl
+++ b/wgpu/examples/cube/shader.wgsl
@@ -23,11 +23,16 @@ fn vs_main(
 [[group(0), binding(1)]]
 var r_color: texture_2d<u32>;
 
+struct PushConstants {
+    mul: f32;
+};
+var<push_constant> pc: PushConstants;
+
 [[stage(fragment)]]
 fn fs_main(in: VertexOutput) -> [[location(0)]] vec4<f32> {
     let tex = textureLoad(r_color, vec2<i32>(in.tex_coord * 256.0), 0);
     let v = f32(tex.x) / 255.0;
-    return vec4<f32>(1.0 - (v * 5.0), 1.0 - (v * 15.0), 1.0 - (v * 50.0), 1.0);
+    return pc.mul * vec4<f32>(1.0 - (v * 5.0), 1.0 - (v * 15.0), 1.0 - (v * 50.0), 1.0);
 }
 
 [[stage(fragment)]]
```

As well as a modified branch of https://gitlab.com/veloren/veloren

**Note**: I don't have much experience with openGL so the code might not be the best.

### CODE REVIEW

**Review by kvark — CHANGES_REQUESTED** [review]
This is a bit unexpected. What I thought we would be doing is - using freestanding uniforms for push constants. That's what gfx-hal did, IIRC. But instead this PR tries to do this via a uniform buffer.

The current approach isn't going to work from the point of view of synchronization between GPU and CPU. Within the same submission, multiple different draw calls would want to overwrite the mapped buffer contents, bind it, and use it on GPU. This means 100% data race between CPU and GPU.

**Comment by JCapucho** [comment]
> This is a bit unexpected. What I thought we would be doing is - using freestanding uniforms for push constants. That's what gfx-hal did, IIRC. But instead this PR tries to do this via a uniform buffer.

Should have probably checked if `gfx-hal` already had this implemented

> The current approach isn't going to work from the point of view of synchronization between GPU and CPU. Within the same submission, multiple different draw calls would want to overwrite the mapped buffer contents, bind it, and use it on GPU. This means 100% data race between CPU and GPU.

Could you elaborate from what I read from the OpenGL wiki about synchronization, https://www.khronos.org/opengl/wiki/Synchronization#Implicit_synchronization

> Most attempts to write to a buffer object, either with glBufferSubData or mapping, will halt until all rendering commands using that part of the buffer object have completed

So unless I'm reading wrong (which is definitely possible), writing to a mapped buffer is implicitly synchronized, even between draw calls

**Comment by kvark** [comment]
First of all, this "will halt" is a terrible thing, and we should not expose a feature that results in that happening potentially many times per frame.
Secondly, it talks about us doing `glMapBuffer`. This is not the same as just writing data into a buffer that's already mapped. With a persistently mapped buffer, the driver can't detect when you are writing to it.

**Comment by JCapucho** [comment]
> First of all, this "will halt" is a terrible thing, and we should not expose a feature that results in that happening potentially many times per frame. Secondly, it talks about us doing `glMapBuffer`. This is not the same as just writing data into a buffer that's already mapped. With a persistently mapped buffer, the driver can't detect when you are writing to it.

Okay, then I'll go with the `gfx-hal` approach, from my understanding the push constants data is copied to a global memory area and then after that it's made into a uniform with [`glUniform`](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glUniform.xhtml) is that correct?

Since `naga` currently only supports one push constant buffer it would mean that it would only support 64 bytes for now

**Comment by kvark** [comment]
> Okay, then I'll go with the gfx-hal approach, from my understanding the push constants data is copied to a global memory area and then after that it's made into a uniform with glUniform is that correct?

Yes. It's not far from what this PR is doing already.

**Comment by JCapucho** [comment]
@kvark I've hit a problem, to use the `glUniform` functions a uniform location needs to be passed, but because `naga` is generating a uniform block it doesn't get a location but a binding, so either `naga` starts outputting non block uniforms for push constants or we can't use this method.

**Comment by JCapucho** [comment]
I've tried fiddling around with the shaders and glsl only allows to use directly the struct without the wrapping block if it also drops the layout qualifier, which isn't that great since then we get implementation defined layouts.

**Comment by kvark** [comment]
Right, naga should not generate them in a uniform block.

**Comment by JCapucho** [comment]
@kvark I've update it to use freestanding uniforms, it's basically a copy of what gfx-hal does.

File: wgpu-hal/src/gles/device.rs
Comment (JCapucho) [inline wgpu-hal/src/gles/device.rs:286]: In my modified buffer example, while not a `sampler2d`, a `usampler2d` triggered this code path, that's why there is a match now to guard against sampler types, because they were being included and causing problems later.

File: Cargo.toml
Comment (JCapucho) [inline Cargo.toml:21]: This is using a patched `naga` for now to allow for iterating, I'll merge it upstream after we converge on a desing and update `naga`

File: wgpu-hal/src/gles/device.rs
Comment (kvark) [reply wgpu-hal/src/gles/device.rs:286]: I wish there was an easier way to detect if this uniform is for a texture. Maybe by name?

File: wgpu-hal/src/gles/mod.rs
Comment (kvark) [inline wgpu-hal/src/gles/mod.rs:722]: let's use `Range<u32>` for consistency with the other things here?

File: wgpu-hal/src/gles/queue.rs
Comment (kvark) [inline wgpu-hal/src/gles/queue.rs:1091]: looks like we could save `Some(&uniform.location.clone())` somewhere for DRY

File: wgpu-hal/src/gles/command.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/command.rs:634]: I copied this straight from gfx-hal but I can see the problem and triggering it is really easy, we could store the size of the uniform and then issue multiple `SetPushConstants` as needed.

File: wgpu-hal/src/gles/device.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/device.rs:286]: By the name would be hard since naga names are like _group_0_binding_0, only option would be changing naga's naming scheme

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [reply wgpu-hal/src/gles/command.rs:636]: yeah, gfx-hal wasn't ideal by any means :)

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [inline wgpu-hal/src/gles/command.rs:167]: We really don't want to clone a heap-allocated container here.
Instead, it should do something like this:
```rust
self.state.uniforms.clear();
self.state.uniforms.extend_from_slice(&inner.uniforms);
```

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [reply wgpu-hal/src/gles/command.rs:636]: I think what we can do here is simply have an array like this:
```rust
push_offset_to_uniform: [Uniform; MAX_PUSH_CONSTANTS];
```
You'd then have some special value (e.g. `!0`) for offsets that don't have a corresponding uniform.
So here, you have a range, and you need to figure out all of the uniforms affected by it. You'd do something simple, like this:
```rust
let mut offset = start_offset;
while offset < end {
  let uniform = &self.state.push_offset_to_uniform[offset as u8];
  self.cmd_buffer
            .commands
            .push(C::SetPushConstants { *uniform, offset..offset+uniform.size });
  offset += uniform.size;
}
```

That's a very straightforward linear scan, it should perform well.

File: wgpu-hal/src/gles/device.rs
Comment (kvark) [inline wgpu-hal/src/gles/device.rs:321]: oh, can we move this match into `conv.rs` so that it doesn't clutter this module too much?

File: wgpu-hal/src/gles/command.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/command.rs:167]: This is using a `Box` so it can't be done this way but I agree with you that cloning a `Box` isn't ideal

Comment (kvark) [reply wgpu-hal/src/gles/command.rs:167]: Can't you switch to a `Vec<>`?

File: wgpu-hal/src/gles/command.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/command.rs:636]: I'm sorry there are some parts that confuse me, first what should be `MAX_PUSH_CONSTANTS` from what I could find there's nothing like it in wgpu yet.

Second how would you assign `0!` if the type of the array is `Uniform`.

Finally, how would this work
```rust
&self.state.push_offset_to_uniform[offset as u8]
```
indexing by `u8` and the index isn't increasing by one but by the size of the previous uniform, which would mean the array should have as many element as the address space of the push constant

**Comment by JCapucho** [comment]
@kvark I've updated the PR with your recommendations, I think the annoying `Vec` copy could be removed if instead we used a `Arc<[UniformDesc; MAX_PUSH_CONSTANTS]>` but this has the downside that it can't derive `Default` so `State` would need a custom impl.

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [inline wgpu-hal/src/gles/command.rs:31]: looks like this is the only part that's heap allocated. This isn't a problem in itself, but we could use `ArrayVec` to solve it

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [inline wgpu-hal/src/gles/command.rs:632]: this should be `offset >> 2` (or `offset / 4`) instead of `offset % 4`

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [inline wgpu-hal/src/gles/command.rs:639]: should this end with `range.start + offset + size` instead of `range.end`?

File: wgpu-hal/src/gles/device.rs
Comment (kvark) [inline wgpu-hal/src/gles/device.rs:293]: similarly, let's move this match into `conv.rs`

**Review by kvark — COMMENTED** [review]
Looks good overall, thank you for doing this!
Just a few smaller corrections are left to do

File: wgpu-hal/src/gles/command.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/command.rs:31]: `ArrayVec` needs for `T` to be `Copy` when using `try_extend_from_slice` and `UniformDesc`doesn't implement it nor can it, since `glow::UniformLocation` isn't `Copy`

File: wgpu-hal/src/gles/command.rs
Comment (JCapucho) [reply wgpu-hal/src/gles/command.rs:639]: Yes, but since the range only specifies where in `data_bytes` is our push constant data, the end functions more like a bound check, so if we leave the end as the end of the allocation it serves the same purpose.

Technically we only need the start offset

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [reply wgpu-hal/src/gles/command.rs:31]: `extend_from_slice` is just a convenience method, which btw isn't requiring `Copy` in std. So let's just `extend()` instead.

File: wgpu-hal/src/gles/command.rs
Comment (kvark) [reply wgpu-hal/src/gles/command.rs:639]: let's remove the end entirely then, no need to carry it around for validation

**Comment by JCapucho** [comment]
@kvark I've finished applying your suggestions, only one problem remains, `UniformLocation` for webgl is not `Sync` which causes `CommandEncoder` to also not be.

**Comment by kvark** [comment]
@JCapucho let's `unsafe impl Sync for UniformDesc`, at least for now

**Comment by JCapucho** [comment]
@kvark everything seems to be working now, deno is failing because it's not picking up the patch for naga but once everything is merged upstream it should be fixed. Are we good to go?

**Review by kvark — APPROVED** [review]
Looking good!
So the last thing here is to change the naga dependency back to upstream.

**Comment by kvark** [comment]
Weird, deno stuff fails but the rest is fine? https://github.com/gfx-rs/wgpu/runs/4900108750?check_suite_focus=true#step:4:255
@crowlKats @lucacasonato

**Comment by JCapucho** [comment]
@kvark I've updated to latest naga, it includes quite some shader changes beacuse of the attribute changes so I've made it into another PR but it can be all squashed

**Comment by crowlKats** [comment]
@kvark seems to me like it was just ci being flaky due to caching

**Review by kvark — APPROVED** [review]
Thank you so much for updating the shaders to the new syntax!

**Pull request merged**
